nanofi/src/core/cstream.c

```diff
@@ -0,0 +1,221 @@
+      } else {
+        if (errno == EINTR) {
+          continue;
+        }
+        logc(err, "Could not recv on %d, error: %s", stream->socket_, strerror(errno));
+      }
+      return -1;
+    }
+    len -= bytes_read;
+    buf += bytes_read;
+    total_read += bytes_read;
+  }
+  if(total_read)
+    logc(trace, "Received data size %d over socket %d", total_read, stream->socket_);
+  return total_read;
+}
+
+int writeUTF(const char * cstr, uint64_t len, enum Bool widen, cstream * stream) {
+  if (len > 65535) {
+    return -1;
+  }
+
+  int ret;
+  if (!widen) {
+    uint16_t shortlen = len;
+    ret = write_uint16_t(shortlen, stream);
+  } else {
+    ret = write_uint32_t(len, stream);
+  }
+
+  if(len == 0 || ret < 0) {
+    return ret;
+  }
+
+  const uint8_t *underlyingPtr = (const uint8_t *)cstr;
+
+  if (!widen) {
+    uint16_t short_length = len;
+    ret = write_buffer(underlyingPtr, short_length, stream);
+  } else {
+    ret = write_buffer(underlyingPtr, len, stream);
+  }
+  return ret;
+}
+
+int read_uint8_t(uint8_t *value, cstream * stream) {
+  uint8_t val;
+  int ret = read_buffer(&val, sizeof(uint8_t), stream);
+  if(ret == sizeof(uint8_t)) {
+    *value = val;
+  }
+  return ret;
+}
+int read_uint16_t(uint16_t *value, cstream * stream) {
+  uint16_t val;
+  int ret = read_buffer((uint8_t*)&val, sizeof(uint16_t), stream);
+  if(ret == sizeof(uint16_t)) {
+    *value = ntohs(val);
+  }
+  return ret;
+}
+int read_uint32_t(uint32_t *value, cstream * stream) {
+  uint32_t val;
+  int ret = read_buffer((uint8_t*)&val, sizeof(uint32_t), stream);
+  if(ret == sizeof(uint32_t)) {
+    *value = ntohl(val);
+  }
+  return ret;
+}
+
+int readUTFLen(uint32_t * utflen, cstream * stream) {
+  int ret = 1;
+  uint16_t shortLength = 0;
+  ret = read_uint16_t(&shortLength, stream);
+  if (ret > 0) {
+    *utflen = shortLength;
+  }
+  return ret;
+}
+
+int readUTF(char * buf, uint64_t buflen, cstream * stream) {
+  //return stream->impl->readData((uint8_t*)buf, buflen);
+  return read_buffer((uint8_t*)buf, buflen, stream);
+}
+
+void close_stream(cstream * stream) {
+  if(stream != NULL && stream->socket_ != -1) {
+#ifdef _WIN32
+    shutdown(stream->socket_, SD_BOTH);
+    closesocket(stream->socket_);
+    WSACleanup();
+#else
+    shutdown(stream->socket_, SHUT_RDWR);
+    close(stream->socket_);
+#endif
+    stream->socket_ = -1;
+  }
+}
+
+cstream * create_socket(const char * host, uint16_t portnum) {
+  logc(trace, "Creating socket to connect to: %s:%d", host, portnum);
+
+#ifdef _WIN32
+  WSADATA wsa;
+  if (WSAStartup(MAKEWORD(2,2),&wsa) != 0)
+	{
+		logc(err, "%s", "WSAStartup failed");
+		return NULL;
+	}
+#endif
+
+  SOCKET sock;
+  struct sockaddr_in server_addr;
+
+  //Create socket
+  sock = socket(AF_INET , SOCK_STREAM , 0);
+  if (sock == -1)
+  {
+    logc(err, "%s", "Failed to create socket");
+    return NULL;
+  }
+
+  struct hostent *server = gethostbyname(host);
```

Comment:
gethostbyname is obsolete and does not support IPv6. As long as we do not need IPv6 support it's okay, but once we do we have to switch to getaddrinfo (and review a few other things).

Reply:
Thanks, modified to be ipv6-compatible.

Reply:
Nice, thank you.